app/services/order_factory.rb

```diff
@@ -0,0 +1,71 @@
+# Builds orders based on a set of attributes
+# There are some idiosyncracies in the order creation process,
+# and it is nice to have them dealt with in one place.
+
+class OrderFactory
+  def initialize(attrs, opts = {})
+    @attrs = attrs.with_indifferent_access
+    @opts = opts.with_indifferent_access
+  end
+
+  def create
+    create_order
+    set_user
+    build_line_items
+    set_addresses
+    create_payment
+    @order
+  end
+
+  private
+
+  attr_reader :attrs, :opts
+
+  def customer
+    @customer ||= Customer.find(attrs[:customer_id])
+  end
+
+  def create_order
+    @order = Spree::Order.create!(create_attrs)
+  end
+
+  def create_attrs
+    create_attrs = attrs.slice(:customer_id, :order_cycle_id, :distributor_id, :shipping_method_id)
+    create_attrs[:email] = customer.email
+    create_attrs
+  end
+
+  def build_line_items
+    attrs[:line_items].each do |li|
+      next unless variant = Spree::Variant.find_by_id(li[:variant_id])
```

Comment:
Is this something we should warn about or is that handled in a later step?

Reply:
Maybe I should just remove the `next unless`? A missing variant here is almost impossible as the relationship between the two tables is described by a foreign key.